src/coalgebras/itreeTauScript.sml

```diff
@@ -1564,6 +1614,1047 @@ Proof
+  >- metis_tac[]
+  >- metis_tac[]
+  >- metis_tac[]
+  >- (imp_res_tac strip_tau_FUNPOW
+      \\ gvs[FUNPOW_Tau_bind]
+      \\ last_x_assum $ qspec_then ‘x’ assume_tac
+      \\ first_assum $ assume_tac o SRULE[Once ret_or_reach_cases]
+      \\ gvs[GSYM FUNPOW_ADD, GSYM ADD_SUC]
+      \\ metis_tac[strip_tau_FUNPOW_strip_tau]
+     )
+  >- (imp_res_tac strip_tau_FUNPOW
+      \\ gvs[FUNPOW_Tau_bind]
+      \\ disj2_tac
+      \\ disj1_tac
+      \\ qexistsl [‘e’, ‘(λx. itree_bind (k' x) k)’] \\ rw[strip_tau_FUNPOW_strip_tau]
+      \\ metis_tac[]
+     )
+  \\ last_assum $ assume_tac o SRULE[Once ret_or_reach_cases]
+  \\ gvs[]
+  >- (imp_res_tac strip_tau_FUNPOW
+      \\ gvs[FUNPOW_Tau_bind, GSYM FUNPOW_ADD, GSYM ADD_SUC]
+      \\ last_x_assum $ qspec_then ‘x’ assume_tac
+      \\ first_assum $ assume_tac o SRULE[Once ret_or_reach_cases]
+      \\ gvs[GSYM FUNPOW_ADD, GSYM ADD_SUC]
+      >- (drule strip_tau_FUNPOW
+          \\ strip_tac
+          \\ gvs[GSYM FUNPOW_ADD, GSYM ADD_SUC]
+          \\ disj1_tac
+          \\ qexists ‘x'’ \\ rw[strip_tau_FUNPOW_strip_tau]
+         )
+      >- metis_tac[strip_tau_FUNPOW_strip_tau]
+      \\ disj1_tac
+      \\ qexists ‘x’ \\ rw[strip_tau_FUNPOW_strip_tau]
+     )
+  >- (imp_res_tac strip_tau_FUNPOW
+      \\ gvs[FUNPOW_Tau_bind, GSYM FUNPOW_ADD, GSYM ADD_SUC]
+      \\ disj2_tac
+      \\ disj1_tac
+      \\ qexistsl [‘e’, ‘(λx. itree_bind (k' x) k)’] \\ rw[strip_tau_FUNPOW_strip_tau]
+      \\ metis_tac[]
+     )
+  \\ drule $ iffLR FUNPOW_Tau_SUC_cyclic_spin
+  \\ metis_tac[spin_bind, spin_FUNPOW_Tau]
+QED
+
+Theorem ret_or_reach_tau:
+  ret_or_reach t'' (Tau t) ==> ret_or_reach t'' t \/ t'' = t
+Proof
+  strip_tac
+  \\ drule $ iffLR ret_or_reach_cases
+  \\ rpt strip_tac
+  >- (gvs[strip_tau_rules]
+      \\ disj1_tac
+      \\ irule $ iffRL ret_or_reach_cases
+      \\ metis_tac[]
+     )
+  >- (gvs[strip_tau_rules]
+      \\ disj1_tac
+      \\ irule $ iffRL ret_or_reach_cases
+      \\ metis_tac[]
+     )
+  \\ gvs[FUNPOW_SUC]
+  \\ Cases_on ‘n’ \\ gvs[FUNPOW]
+  \\ disj1_tac
+  \\ irule $ iffRL ret_or_reach_cases
+  \\ metis_tac[GSYM FUNPOW]
+QED
+
+Theorem ret_or_reach_vis:
+  ret_or_reach t (Vis e k) ==> !l. ret_or_reach t (k l) \/ t = k l
+Proof
+  rpt strip_tac
+  \\ drule $ iffLR ret_or_reach_cases
+  \\ gvs[strip_tau_rules, FUNPOW_SUC]
+  \\ metis_tac[]
+QED
+
+Theorem ret_or_reach_funpow_vis:
+  ret_or_reach t (FUNPOW Tau n (Vis e k)) ==> (!l. ret_or_reach t (k l) \/ t = k l)
+                                              \/ (?n'. FUNPOW Tau n (Vis e k) = FUNPOW Tau (SUC n') t)
+Proof
+  rpt strip_tac
+  \\ drule $ iffLR ret_or_reach_cases
+  \\ rw[]
+  >- (imp_res_tac strip_tau_FUNPOW
+      \\ gvs[FUNPOW_Tau_neq2]
+     )
+  >- (imp_res_tac strip_tau_FUNPOW
+      \\ imp_res_tac FUNPOW_Tau_Vis_eq
+      \\ metis_tac[]
+     )
+  \\ gvs[FUNPOW_SUC]
+  \\ metis_tac[]
+QED
+
+(* return or reaching an instance of an abstracted function *)
+CoInductive ret_or_reach_abs:
```

Comment:
I would find this definition more natural if `abs` was a set of trees instead of a `ktree`. The arguments to `abs` are, after all, never relevant for your purposes; it only seems to matter whether a particular tree is in the domain or `abs` or not. Sets are a better fit for this.